<a href="https://colab.research.google.com/github/JainAnki/SiameseNetworks/blob/main/5a_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Installing pandas-profiling library

In [2]:
! pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip

  Using cached https://github.com/pandas-profiling/pandas-profiling/archive/master.zip


#Importing Libraries

In [3]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import pandas_profiling as pp
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

In [4]:
#%matplotlib inline

import matplotlib
# Force matplotlib to not use any Xwindows backend.
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(color_codes=True)
from scipy import stats

#import plotly.plotly as py

import plotly.graph_objs as go
from plotly.offline import init_notebook_mode,iplot
#init_notebook_mode(connected=True)

#Reading the "loan_train.csv" file
###1.   Read .csv file
###2.   Print the info of the dataframe
###3.   Print the top 5 rows of the dataframe



In [6]:
df = pd.read_csv('/content/loan_train.csv', low_memory=False)
df = df[(df['loan_status'] == 'Fully Paid') | (df['loan_status'] == 'Charged Off')]
print("df.info()")
print(df.info())
print("\n")
print("df.head()")
print(df.head())
print("\n")
print(df.shape)

df.info()
<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 20.8+ MB
None


df.head()
       id  member_id  ...  total_bc_limit  total_il_high_credit_limit
0  573354     737474  ...             NaN                         NaN
1  476321     603324  ...             NaN                         NaN
2  451484     556265  ...             NaN                         NaN
4  800018    1005270  ...             NaN                         NaN
5  471391     595223  ...             NaN                         NaN

[5 rows x 111 columns]


(24301, 111)


#Creating and saving the ProfileReport of the **original** dataframe 



In [7]:
profile = pp.ProfileReport(df, title = 'Loan_dataset', html = {'style': {'full_width': True}})
profile.to_notebook_iframe()
profile.to_file(output_file = 'Loan_dataset.html')


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

#Reading the "loan_train.csv" file
###1.   Read .csv file to perform exploratory data analysis (EDA)
###2.   From the ProfileReport it was seen that a number of attributes had


*   only NaN values
*   a single value for all the data observations
  ### So we remove these attributes from the dataframe




In [34]:
df_re = pd.read_csv('/content/loan_train.csv', low_memory=False)
df_re = df_re[(df_re['loan_status'] == 'Fully Paid') | (df_re['loan_status'] == 'Charged Off')]
mapping_dict = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_re = df_re.replace(mapping_dict)
df_re.loan_status.value_counts()
df_re = df_re.drop(['next_pymnt_d','id','member_id','pymnt_plan','url','initial_list_status','mths_since_last_major_derog','policy_code','application_type','annual_inc_joint','dti_joint','verification_status_joint','acc_now_delinq','tot_coll_amt','tot_cur_bal','open_acc_6m','open_il_6m','open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il','il_util',	'open_rv_12m','open_rv_24m','max_bal_bc','all_util','total_rev_hi_lim', 'inq_fi','total_cu_tl','inq_last_12m','acc_open_past_24mths','avg_cur_bal','bc_open_to_buy','bc_util','delinq_amnt','mo_sin_old_il_acct','mo_sin_old_rev_tl_op','mo_sin_rcnt_rev_tl_op','mo_sin_rcnt_tl','mort_acc','mths_since_recent_bc','mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd','num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats','num_bc_tl','num_il_tl','num_op_rev_tl','num_rev_accts','num_rev_tl_bal_gt_0',		'num_sats',	'num_tl_120dpd_2m','num_tl_30dpd','num_tl_90g_dpd_24m',	'num_tl_op_past_12m',	'pct_tl_nvr_dlq','percent_bc_gt_75','tot_hi_cred_lim','total_bal_ex_mort','total_bc_limit',	'total_il_high_credit_limit'], axis =1)

df_re_te = pd.read_csv('/content/loan_test.csv', low_memory=False)
df_re_te = df_re_te[(df_re_te['loan_status'] == 'Fully Paid') | (df_re_te['loan_status'] == 'Charged Off')]
mapping_dict = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_re_te = df_re_te.replace(mapping_dict)
df_re_te.loan_status.value_counts()
df_re_te = df_re_te.drop(['next_pymnt_d','id','member_id','pymnt_plan','url','initial_list_status','mths_since_last_major_derog','policy_code','application_type','annual_inc_joint','dti_joint','verification_status_joint','acc_now_delinq','tot_coll_amt','tot_cur_bal','open_acc_6m','open_il_6m','open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il','il_util',	'open_rv_12m','open_rv_24m','max_bal_bc','all_util','total_rev_hi_lim', 'inq_fi','total_cu_tl','inq_last_12m','acc_open_past_24mths','avg_cur_bal','bc_open_to_buy','bc_util','delinq_amnt','mo_sin_old_il_acct','mo_sin_old_rev_tl_op','mo_sin_rcnt_rev_tl_op','mo_sin_rcnt_tl','mort_acc','mths_since_recent_bc','mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd','num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats','num_bc_tl','num_il_tl','num_op_rev_tl','num_rev_accts','num_rev_tl_bal_gt_0',		'num_sats',	'num_tl_120dpd_2m','num_tl_30dpd','num_tl_90g_dpd_24m',	'num_tl_op_past_12m',	'pct_tl_nvr_dlq','percent_bc_gt_75','tot_hi_cred_lim','total_bal_ex_mort','total_bc_limit',	'total_il_high_credit_limit'], axis =1)



##From the ProfileReport it was also seen that some of the attributes have
*   NaN + missing values as can be seen below

So these attributes were also removed and the number of attributes have now reduced from 111 to 45.





In [ ]:
print('df_re[collections_12_mths_ex_med].unique()')
print(df_re['collections_12_mths_ex_med'].unique())
print('\n')
print('df_re[chargeoff_within_12_mths].unique()')
print(df_re['chargeoff_within_12_mths'].unique())
print('\n')
print('df_re[tax_liens].unique()')
print(df_re['tax_liens'].unique())
df_re = df_re.drop(['collections_12_mths_ex_med','chargeoff_within_12_mths','tax_liens'], axis= 1)
print('\n')
print('df_re.shape = ', df_re.shape)

print('df_re_te[collections_12_mths_ex_med].unique()')
print(df_re_te['collections_12_mths_ex_med'].unique())
print('\n')
print('df_re_te[chargeoff_within_12_mths].unique()')
print(df_re_te['chargeoff_within_12_mths'].unique())
print('\n')
print('df_re_te[tax_liens].unique()')
print(df_re_te['tax_liens'].unique())
df_re_te = df_re_te.drop(['collections_12_mths_ex_med','chargeoff_within_12_mths','tax_liens'], axis= 1)
print('\n')
print('df_re_te.shape = ', df_re_te.shape)

#Creating and saving the ProfileReport of the **reduced** dataframe 


In [ ]:
profile_re = pp.ProfileReport(df_re, title = 'Loan_dataset', html = {'style': {'full_width': True}})
profile_re.to_notebook_iframe()
profile_re.to_file(output_file = 'Loan_dataset_re.html')

In [ ]:
df_re.to_csv('cleaned_dataset.csv',index=False)
df_re.shape

#Using python to reduce the dataset instead of analyzing ProfileReport

In [5]:
df_co = pd.read_csv('/content/loan_train.csv', low_memory=False)
df_co_te = pd.read_csv('/content/loan_test.csv', low_memory=False)

In [6]:
print(df_co.isnull().sum())
df_co = df_co[(df_co['loan_status'] == 'Fully Paid') | (df_co['loan_status'] == 'Charged Off')]
mapping_dict = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_co = df_co.replace(mapping_dict)
print("\n")
print(df_co_te.isnull().sum())
df_co_te = df_co_te[(df_co_te['loan_status'] == 'Fully Paid') | (df_co_te['loan_status'] == 'Charged Off')]
mapping_dict_te = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_co_te = df_co_te.replace(mapping_dict_te)


id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        17
tot_hi_cred_lim               24999
total_bal_ex_mort             24999
total_bc_limit                24999
total_il_high_credit_limit    24999
Length: 111, dtype: int64


id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        22
tot_hi_cred_lim               14718
total_bal_ex_mort             14718
total_bc_limit                14718
total_il_high_credit_limit    14718
Length: 111, dtype: int64


In [7]:

drop_columns = []
for col in df_co.columns:
    non_null = df_co[col].dropna()
    unique_non_null = non_null.unique()
    num_true_unique = len(unique_non_null)
    if num_true_unique < 2:
        drop_columns.append(col)
df_co = df_co.drop(drop_columns, axis=1)

drop_columns_te = []
for col in df_co_te.columns:
    non_null = df_co_te[col].dropna()
    unique_non_null = non_null.unique()
    num_true_unique = len(unique_non_null)
    if num_true_unique < 2:
        drop_columns_te.append(col)
df_co_te = df_co_te.drop(drop_columns, axis=1)

In [8]:
display(df_co.shape,
df_co_te.shape)

(24301, 45)

(14276, 45)

# Drop any column with more than 50% missing values


In [9]:
half_count = len(df_co) / 2
df_co = df_co.dropna(thresh=half_count,axis=1)

half_count = len(df_co_te) / 2
df_co_te = df_co_te.dropna(thresh=half_count,axis=1)


In [10]:
display(df_co.shape,
df_co_te.shape)

(24301, 43)

(14276, 43)

# Checking which attributes are of 'object' DataType

In [11]:
display(df_co.select_dtypes(include=['object']).head(5),
df_co_te.select_dtypes(include=['object']).head(5))

term int_rate grade  ... revol_util last_pymnt_d last_credit_pull_d
0   36 months    7.51%     A  ...      1.20%       11-Mar             11-Jul
1   36 months    8.94%     A  ...        14%       10-Mar             10-Feb
2   36 months   13.57%     C  ...     34.40%       11-Dec             11-Dec
4   60 months   17.49%     D  ...      9.40%       15-Sep             16-May
5   36 months    8.94%     A  ...      2.80%       10-May             10-May

[5 rows x 19 columns]

term int_rate grade  ... revol_util last_pymnt_d last_credit_pull_d
0   36 months    6.62%     A  ...     11.60%       14-Nov             16-May
1   36 months   13.49%     C  ...     11.90%       14-May             14-Sep
2   36 months    8.94%     A  ...     46.40%       10-Feb             16-May
3   60 months   16.49%     D  ...     60.20%       16-May             16-May
4   36 months   15.21%     D  ...     55.70%       13-Sep             16-May

[5 rows x 19 columns]

#Transforming 'objects' with 'float' values to 'float' data type
##1. 'int_rate'
##2. 'revol_util'
##3. 'term'
##4. 'emp_length'

In [12]:
df_co['int_rate'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co['int_rate'] = df_co['int_rate'].astype(float)

df_co['revol_util'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co['revol_util'] = df_co['revol_util'].astype(float)

df_co['term'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co['term'] = df_co['term'].astype(int)
set(df_co["term"])

#display(set(df_co["emp_length"]))
df_co.replace('n/a', np.nan,inplace=True)
df_co.emp_length.fillna(value=0,inplace=True)
set(df_co["emp_length"])
df_co['emp_length'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co['emp_length'] = df_co['emp_length'].astype(int)
set(df_co["emp_length"])


df_co_te['int_rate'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co_te['int_rate'] = df_co_te['int_rate'].astype(float)

df_co_te['revol_util'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co_te['revol_util'] = df_co_te['revol_util'].astype(float)

df_co_te['term'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co_te['term'] = df_co_te['term'].astype(int)
set(df_co_te["term"])

#display(set(df_co["emp_length"]))
df_co_te.replace('n/a', np.nan,inplace=True)
df_co_te.emp_length.fillna(value=0,inplace=True)
set(df_co_te["emp_length"])
df_co_te['emp_length'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co_te['emp_length'] = df_co_te['emp_length'].astype(int)
set(df_co_te["emp_length"])

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

In [13]:
display(df_co.select_dtypes(include=['object']).head(5),
df_co_te.select_dtypes(include=['object']).head(5))

grade sub_grade  ... last_pymnt_d last_credit_pull_d
0     A        A4  ...       11-Mar             11-Jul
1     A        A5  ...       10-Mar             10-Feb
2     C        C3  ...       11-Dec             11-Dec
4     D        D5  ...       15-Sep             16-May
5     A        A5  ...       10-May             10-May

[5 rows x 15 columns]

grade sub_grade  ... last_pymnt_d last_credit_pull_d
0     A        A2  ...       14-Nov             16-May
1     C        C2  ...       14-May             14-Sep
2     A        A5  ...       10-Feb             16-May
3     D        D3  ...       16-May             16-May
4     D        D2  ...       13-Sep             16-May

[5 rows x 15 columns]

#Transforming Categorical values to 'str' data type
##1. grade
##2. sub_grade
##3. home_ownership
##4. verification_status

In [14]:
print(set(df_co['grade']),"\n",
set(df_co['sub_grade']),"\n",
set(df_co['home_ownership']),"\n",
set(df_co['verification_status']),"\n",
set(df_co['purpose']))
df_co['grade']= df_co['grade'].convert_dtypes()
df_co['sub_grade']=df_co['sub_grade'].convert_dtypes()
df_co['home_ownership'] = df_co['home_ownership'].convert_dtypes()
df_co['verification_status'] = df_co['verification_status'].convert_dtypes()
df_co['purpose']=df_co['purpose'].convert_dtypes()

print(set(df_co_te['grade']),"\n",
set(df_co_te['sub_grade']),"\n",
set(df_co_te['home_ownership']),"\n",
set(df_co_te['verification_status']),"\n",
set(df_co_te['purpose']))
df_co_te['grade']= df_co_te['grade'].convert_dtypes()
df_co_te['sub_grade']=df_co_te['sub_grade'].convert_dtypes()
df_co_te['home_ownership'] = df_co_te['home_ownership'].convert_dtypes()
df_co_te['verification_status'] = df_co_te['verification_status'].convert_dtypes()
df_co_te['purpose']=df_co['purpose'].convert_dtypes()

display(df_co.select_dtypes(include=['object']).head(5),
df_co_te.select_dtypes(include=['object']).head(5))

{'A', 'D', 'B', 'C', 'E', 'G', 'F'} 
 {'G4', 'E2', 'B3', 'B1', 'D2', 'A5', 'F3', 'B2', 'A3', 'F5', 'A4', 'F2', 'C5', 'F4', 'E4', 'C2', 'E3', 'G3', 'A2', 'G5', 'A1', 'B5', 'F1', 'G2', 'D3', 'E5', 'D4', 'E1', 'B4', 'C3', 'C4', 'D5', 'D1', 'C1', 'G1'} 
 {'OTHER', 'MORTGAGE', 'RENT', 'OWN'} 
 {'Not Verified', 'Verified', 'Source Verified'} 
 {'medical', 'vacation', 'home_improvement', 'small_business', 'moving', 'car', 'renewable_energy', 'wedding', 'house', 'educational', 'other', 'credit_card', 'major_purchase', 'debt_consolidation'}
{'A', 'D', 'B', 'C', 'E', 'G', 'F'} 
 {'G4', 'E2', 'B3', 'B1', 'D2', 'A5', 'F3', 'B2', 'A3', 'A4', 'E4', 'F2', 'C5', 'F4', 'F5', 'C2', 'E3', 'G3', 'A2', 'G5', 'A1', 'B5', 'F1', 'D3', 'G2', 'E5', 'D4', 'E1', 'B4', 'C4', 'C3', 'D5', 'D1', 'C1', 'G1'} 
 {'OWN', 'NONE', 'MORTGAGE', 'RENT', 'OTHER'} 
 {'Not Verified', 'Verified', 'Source Verified'} 
 {'medical', 'vacation', 'home_improvement', 'moving', 'car', 'small_business', 'renewable_energy', 'wedding', 'hou

emp_title issue_d  ... last_pymnt_d last_credit_pull_d
0                       15  10-Sep  ...       11-Mar             11-Jul
1                     1400  10-Jan  ...       10-Mar             10-Feb
2                    36000   9-Oct  ...       11-Dec             11-Dec
4             old palm inc  11-Jul  ...       15-Sep             16-May
5   Brocade Communications   9-Dec  ...       10-May             10-May

[5 rows x 10 columns]

emp_title issue_d  ... last_pymnt_d last_credit_pull_d
0     PPG Industries, Inc.  11-Nov  ...       14-Nov             16-May
1  PPI Technology Services  11-Aug  ...       14-May             14-Sep
2          PPL Corporation   9-Sep  ...       10-Feb             16-May
3                   ppl eu  11-May  ...       16-May             16-May
4                     PPMM  10-Sep  ...       13-Sep             16-May

[5 rows x 10 columns]

#Removing missing values

In [15]:
print(df_co.isnull().sum())
print(df_co_te.isnull().sum())

id                            0
member_id                     0
loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
emp_title                     6
emp_length                    0
home_ownership                0
annual_inc                    0
verification_status           0
issue_d                       0
loan_status                   0
url                           0
desc                       7943
purpose                       0
title                         8
zip_code                      0
addr_state                    0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                   29
total_ac

In [16]:
df_co = df_co.fillna(df_co['emp_title'].value_counts().index[0])
for col in ['emp_title','issue_d','desc','title', 'zip_code','addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  df_co[col] = df_co[col].astype('category')
print(df_co.dtypes)

df_co_te = df_co_te.fillna(df_co_te['emp_title'].value_counts().index[0])
for col in ['emp_title','issue_d','desc','title', 'zip_code','addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  df_co_te[col] = df_co_te[col].astype('category')
print(df_co_te.dtypes)

id                            int64
member_id                     int64
loan_amnt                     int64
funded_amnt                   int64
funded_amnt_inv             float64
term                          int64
int_rate                    float64
installment                 float64
grade                        string
sub_grade                    string
emp_title                  category
emp_length                    int64
home_ownership               string
annual_inc                  float64
verification_status          string
issue_d                    category
loan_status                   int64
url                          object
desc                       category
purpose                      string
title                      category
zip_code                   category
addr_state                 category
dti                         float64
delinq_2yrs                   int64
earliest_cr_line           category
inq_last_6mths                int64
open_acc                    

#Label Encoding categorical data
###1.grade
###2.sub_grade
###3.emp_title
###4.issue_d
###5.desc
###6.title
###7.addr_state
###8.earliest_cr_line
###9.last_pymnt_d
###10.revol_util
###11.last_pymnt_amnt
###12.last_credit_pull_d
###13.pub_rec_bankruptcies

In [17]:
df_co1 = df_co.select_dtypes(include=['object']).copy()
for col in ['grade','sub_grade','emp_title','issue_d','desc','title', 'addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  le = LabelEncoder()
  df_co[col] = le.fit_transform(df_co[col].astype(str))

df_co_te1 = df_co_te.select_dtypes(include=['object']).copy()
for col in ['grade','sub_grade','emp_title','issue_d','desc','title', 'addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  le = LabelEncoder()
  df_co_te[col] = le.fit_transform(df_co_te[col].astype(str))


In [18]:
df_co.dtypes
df_co_te.dtypes

id                            int64
member_id                     int64
loan_amnt                     int64
funded_amnt                   int64
funded_amnt_inv             float64
term                          int64
int_rate                    float64
installment                 float64
grade                         int64
sub_grade                     int64
emp_title                     int64
emp_length                    int64
home_ownership               string
annual_inc                  float64
verification_status          string
issue_d                       int64
loan_status                   int64
url                          object
desc                          int64
purpose                      string
title                         int64
zip_code                   category
addr_state                    int64
dti                         float64
delinq_2yrs                   int64
earliest_cr_line              int64
inq_last_6mths                int64
open_acc                    

#One Hot Encoding for nominal attributes
##1. purpose
##3. home_ownership
##4. verification_status

In [19]:
nom_cols = ['home_ownership','verification_status','purpose']
df_du = pd.get_dummies(df_co[nom_cols])
df_co = pd.concat([df_co, df_du], axis=1)
df_co = df_co.drop(nom_cols, axis=1)
df_co.head()

nom_cols1 = ['home_ownership','verification_status','purpose']
df_du1 = pd.get_dummies(df_co_te[nom_cols1])
df_co_te = pd.concat([df_co_te, df_du1], axis=1)

df_co_te = df_co_te.drop(nom_cols1, axis=1)
df_co_te.head()

id  member_id  ...  last_credit_pull_d  pub_rec_bankruptcies
0  1001205    1227186  ...                  76                     0
1   850106    1061893  ...                  59                     0
2   444243     541365  ...                  76                     0
3   741007     938805  ...                  76                     0
4   572468     736352  ...                  76                     0

[5 rows x 40 columns]

#Describing data and checking if any more missing values  exist

In [20]:
display(df_co.describe(),
df_co.dtypes.value_counts())
obj = df_co.columns[df_co.dtypes == 'object']
df_co[obj].isnull().sum()

display(df_co_te.describe(),
df_co_te.dtypes.value_counts())
obj = df_co_te.columns[df_co_te.dtypes == 'object']
df_co_te[obj].isnull().sum()

id     member_id  ...  last_credit_pull_d  pub_rec_bankruptcies
count  2.430100e+04  2.430100e+04  ...        24301.000000          24301.000000
mean   6.759235e+05  8.419130e+05  ...           56.448212              0.092712
std    2.080452e+05  2.630066e+05  ...           20.133192              0.432911
min    5.574200e+04  7.367300e+04  ...            0.000000              0.000000
25%    5.126710e+05  6.619030e+05  ...           42.000000              0.000000
50%    6.557640e+05  8.382220e+05  ...           62.000000              0.000000
75%    8.269870e+05  1.035246e+06  ...           76.000000              0.000000
max    1.076863e+06  1.304884e+06  ...          101.000000              3.000000

[8 rows x 38 columns]

int64       26
float64     12
object       1
category     1
dtype: int64

id     member_id  ...  last_credit_pull_d  pub_rec_bankruptcies
count  1.427600e+04  1.427600e+04  ...        14276.000000          14276.000000
mean   6.771536e+05  8.429164e+05  ...           55.881549              0.104021
std    2.113273e+05  2.669023e+05  ...           20.346079              0.459386
min    5.473400e+04  7.069900e+04  ...            0.000000              0.000000
25%    5.112320e+05  6.595262e+05  ...           41.000000              0.000000
50%    6.577305e+05  8.407310e+05  ...           60.000000              0.000000
75%    8.318285e+05  1.040137e+06  ...           75.000000              0.000000
max    1.077501e+06  1.314167e+06  ...          103.000000              3.000000

[8 rows x 38 columns]

int64       26
float64     12
object       1
category     1
dtype: int64

url    0
dtype: int64

In [21]:
df_co.shape
df_co_te.shape

(14276, 40)

#Encoding zipcode

In [22]:
df_co['zip_code'].unique()
df_co['zip_code'] = df_co['zip_code'].str.rstrip('x')
print(df_co['zip_code'].head())

df_co_te['zip_code'].unique()
df_co_te['zip_code'] = df_co_te['zip_code'].str.rstrip('x')
print(df_co_te['zip_code'].head())

df_co['zip_code'] = df_co['zip_code'].astype(int)
df_co_te['zip_code'] = df_co_te['zip_code'].astype(int)

0    076
1    300
2    310
4    334
5    939
Name: zip_code, dtype: object
0    152
1    705
2    181
3    187
4    950
Name: zip_code, dtype: object


In [23]:
def zip_formatted(row):
    if len(str(row)) == 1:
        row = str(row)+ '00'
    elif len(str(row)) == 2:
        row = str(row)+ '0'
    return int(row)

In [24]:
df_co['zip_code'] = df_co['zip_code'].apply(zip_formatted)
df_co_te['zip_code'] = df_co_te['zip_code'].apply(zip_formatted)

In [25]:
min_value = df_co['loan_amnt'].min()
max_value = df_co['loan_amnt'].max()
print(min_value)
print(max_value)
bins = np.linspace(min_value,max_value,5)
print(len(bins))
labels = ['0','1','2','3']
df_co['bins'] = pd.cut(df_co['loan_amnt'], bins=bins, labels=labels, include_lowest=True)
df_co['bins']

min_value = df_co_te['loan_amnt'].min()
max_value = df_co_te['loan_amnt'].max()
print(min_value)
print(max_value)
bins = np.linspace(min_value,max_value,5)
print(len(bins))
labels = ['0','1','2','3']
df_co_te['bins'] = pd.cut(df_co_te['loan_amnt'], bins=bins, labels=labels, include_lowest=True)
df_co_te['bins']

500
35000
5
500
35000
5


0        0
1        0
2        1
3        0
4        2
        ..
14713    0
14714    0
14715    0
14716    0
14717    0
Name: bins, Length: 14276, dtype: category
Categories (4, object): ['0' < '1' < '2' < '3']

#Variable to check whether loan will default

In [26]:
bad_indicator = df_co["loan_status"].isin(["1"])
bad_indicator.value_counts()

bad_indicator_te = df_co_te["loan_status"].isin(["1"])
bad_indicator_te.value_counts()


True     12123
False     2153
Name: loan_status, dtype: int64

#Checking correlation value of loan_status with other attributes

In [27]:
X_tr= df_co.copy()
X_te = df_co_te.copy()
a = df_co.corrwith(df_co['loan_status'])
a

id                        -0.048124
member_id                 -0.044758
loan_amnt                 -0.058606
funded_amnt               -0.056209
funded_amnt_inv           -0.038255
term                      -0.169273
int_rate                  -0.210081
installment               -0.026918
grade                     -0.199933
sub_grade                 -0.202964
emp_title                 -0.028804
emp_length                -0.011834
annual_inc                 0.040116
issue_d                   -0.008817
loan_status                1.000000
desc                      -0.017400
title                     -0.047286
zip_code                  -0.020531
addr_state                 0.021916
dti                       -0.046135
delinq_2yrs               -0.013968
earliest_cr_line           0.013776
inq_last_6mths            -0.075122
open_acc                   0.000698
pub_rec                   -0.050736
revol_bal                 -0.006231
revol_util                -0.086858
total_acc                  0

#Removing attributes with negative correlation values

In [28]:
X_tr=X_tr.drop(['loan_amnt','funded_amnt','funded_amnt_inv','term','int_rate','installment','grade','sub_grade'],axis =1)
X_tr = X_tr.drop(['emp_title','emp_length','issue_d','title','zip_code','addr_state','dti','delinq_2yrs'],axis = 1)
X_tr = X_tr.drop(['earliest_cr_line','inq_last_6mths','pub_rec','revol_bal','revol_util','last_pymnt_amnt','pub_rec_bankruptcies'],axis = 1)
X_tr= X_tr.drop(['total_rec_int','total_rec_late_fee','recoveries','collection_recovery_fee'],axis =1)
X_tr = X_tr.drop(['id','member_id','url'], axis =1)

X_te=X_te.drop(['loan_amnt','funded_amnt','funded_amnt_inv','term','int_rate','installment','grade','sub_grade'],axis =1)
X_te = X_te.drop(['emp_title','emp_length','issue_d','title','zip_code','addr_state','dti','delinq_2yrs'],axis = 1)
X_te = X_te.drop(['earliest_cr_line','inq_last_6mths','pub_rec','revol_bal','revol_util','last_pymnt_amnt','pub_rec_bankruptcies'],axis = 1)
X_te= X_te.drop(['total_rec_int','total_rec_late_fee','recoveries','collection_recovery_fee'],axis =1)
X_te = X_te.drop(['id','member_id','url'], axis =1)

In [29]:
X_tr.to_csv('filtered_loans_train.csv',index=False)
X_tr.head()

X_te.to_csv('filtered_loans_test.csv',index=False)
X_te.head()

annual_inc  loan_status  desc  ...  last_pymnt_d  last_credit_pull_d  bins
0     69000.0            1  9377  ...            57                  76     0
1     24000.0            0  4175  ...            56                  59     0
2     75000.0            1  8850  ...             3                  76     1
3     60000.0            1  9377  ...            76                  76     0
4    135000.0            1  9377  ...            47                  76     2

[5 rows x 11 columns]

In [30]:
X_tr['loan_status'].value_counts() / X_tr.shape[0]

1    0.857043
0    0.142957
Name: loan_status, dtype: float64

#Training Gradirent Boosting Classifier

In [34]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import cross_val_score, GridSearchCV
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.metrics import classification_report

#Creating traian and test data and labels

In [32]:
x_cols = X_tr.columns
features_cols = x_cols.drop(['loan_status'])
y_tr = X_tr['loan_status']
X_tr = X_tr[features_cols]

x_cols = X_te.columns
features_cols = x_cols.drop(['loan_status'])
y_te = X_te['loan_status']
X_te = X_te[features_cols]

#5a

#1st iteration of Grid search

In [118]:
grid = dict()
grid['n_estimators'] = [10, 50, 100, 500]
grid['learning_rate'] = [0.01, 0.1]
grid['subsample'] = [0.5, 0.9]
scores = ["precision", "recall"]
# define the model with default hyperparameters
model = GradientBoostingClassifier()
# define the evaluation procedure
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

# define the grid search procedure
for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring="%s_macro" % score)
    # execute the grid search
    grid_result = grid_search.fit(X_tr, y_tr)
    # summarize the best score and configuration
    print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
    # summarize all scores that were evaluated
    print("Grid scores on development set:")
    print()
    means = grid_result.cv_results_['mean_test_score']
    stds = grid_result.cv_results_['std_test_score']
    params = grid_result.cv_results_['params']
    for mean, stdev, param in zip(means, stds, params):
      print("%f (%f) with: %r" % (mean, stdev, param))
    print()

    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    y_true, y_pred = y_te, grid_search.predict(X_te)
    print(classification_report(y_true, y_pred))
    print()

# Tuning hyper-parameters for precision

Best: 0.973892 using {'learning_rate': 0.1, 'n_estimators': 500, 'subsample': 0.9}
Grid scores on development set:

0.428521 (0.000098) with: {'learning_rate': 0.01, 'n_estimators': 10, 'subsample': 0.5}
0.428521 (0.000098) with: {'learning_rate': 0.01, 'n_estimators': 10, 'subsample': 0.9}
0.930774 (0.000808) with: {'learning_rate': 0.01, 'n_estimators': 50, 'subsample': 0.5}
0.930359 (0.000767) with: {'learning_rate': 0.01, 'n_estimators': 50, 'subsample': 0.9}
0.955575 (0.004276) with: {'learning_rate': 0.01, 'n_estimators': 100, 'subsample': 0.5}
0.955358 (0.004399) with: {'learning_rate': 0.01, 'n_estimators': 100, 'subsample': 0.9}
0.969824 (0.004820) with: {'learning_rate': 0.01, 'n_estimators': 500, 'subsample': 0.5}
0.968788 (0.004597) with: {'learning_rate': 0.01, 'n_estimators': 500, 'subsample': 0.9}
0.955838 (0.004221) with: {'learning_rate': 0.1, 'n_estimators': 10, 'subsample': 0.5}
0.955385 (0.004732) with: {'learning_rate': 0.1,

In [119]:
grid_result

GridSearchCV(cv=RepeatedStratifiedKFold(n_repeats=3, n_splits=10, random_state=1),
             error_score=nan,
             estimator=GradientBoostingClassifier(ccp_alpha=0.0,
                                                  criterion='friedman_mse',
                                                  init=None, learning_rate=0.1,
                                                  loss='deviance', max_depth=3,
                                                  max_features=None,
                                                  max_leaf_nodes=None,
                                                  min_impurity_decrease=0.0,
                                                  min_impurity_split=None,
                                                  min_samples_leaf=1,
                                                  min_samples_spl...
                                                  n_iter_no_change=None,
                                                  presort='deprecated',
          

#2nd run of Grid Search

In [122]:
grid = dict()
grid['n_estimators'] = [100, 500, 750, 1000]
grid['learning_rate'] = [0.001, 1]
grid['subsample'] = [0.5, 0.9]
scores = ["precision", "recall"]
# define the model with default hyperparameters
model_1 = GradientBoostingClassifier()
# define the evaluation procedure
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

# define the grid search procedure
for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring="%s_macro" % score)
    # execute the grid search
    grid_result = grid_search.fit(X_tr, y_tr)
    # summarize the best score and configuration
    print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
    # summarize all scores that were evaluated
    print("Grid scores on development set:")
    print()
    means = grid_result.cv_results_['mean_test_score']
    stds = grid_result.cv_results_['std_test_score']
    params = grid_result.cv_results_['params']
    for mean, stdev, param in zip(means, stds, params):
      print("%f (%f) with: %r" % (mean, stdev, param))
    print()

    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    y_true, y_pred = y_te, grid_search.predict(X_te)
    print(classification_report(y_true, y_pred))
    print()

# Tuning hyper-parameters for precision

Best: 0.955529 using {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.5}
Grid scores on development set:

0.428521 (0.000098) with: {'learning_rate': 0.001, 'n_estimators': 100, 'subsample': 0.5}
0.428521 (0.000098) with: {'learning_rate': 0.001, 'n_estimators': 100, 'subsample': 0.9}
0.930607 (0.000481) with: {'learning_rate': 0.001, 'n_estimators': 500, 'subsample': 0.5}
0.880353 (0.150654) with: {'learning_rate': 0.001, 'n_estimators': 500, 'subsample': 0.9}
0.954594 (0.003399) with: {'learning_rate': 0.001, 'n_estimators': 750, 'subsample': 0.5}
0.954913 (0.003327) with: {'learning_rate': 0.001, 'n_estimators': 750, 'subsample': 0.9}
0.955529 (0.004093) with: {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.5}
0.955381 (0.004392) with: {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.9}
0.868581 (0.101563) with: {'learning_rate': 1, 'n_estimators': 100, 'subsample': 0.5}
0.904524 (0.085140) with: {'lear

In [123]:
grid_result

GridSearchCV(cv=RepeatedStratifiedKFold(n_repeats=3, n_splits=10, random_state=1),
             error_score=nan,
             estimator=GradientBoostingClassifier(ccp_alpha=0.0,
                                                  criterion='friedman_mse',
                                                  init=None, learning_rate=0.1,
                                                  loss='deviance', max_depth=3,
                                                  max_features=None,
                                                  max_leaf_nodes=None,
                                                  min_impurity_decrease=0.0,
                                                  min_impurity_split=None,
                                                  min_samples_leaf=1,
                                                  min_samples_spl...
                                                  n_iter_no_change=None,
                                                  presort='deprecated',
          

#3rd run of Grid Search

In [124]:
grid = dict()
grid['n_estimators'] = [ 500]
grid['learning_rate'] = [0.01]
grid['subsample'] = [0.9]
scores = ["precision","recall"]
# define the model with default hyperparameters
model = GradientBoostingClassifier()
# define the evaluation procedure
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

# define the grid search procedure
for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    grid_search_5a = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring="%s_macro" % score)
    # execute the grid search
    grid_result_5a = grid_search.fit(X_tr, y_tr)
    # summarize the best score and configuration
    print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
    # summarize all scores that were evaluated
    print("Grid scores on development set:")
    print()
    means = grid_result.cv_results_['mean_test_score']
    stds = grid_result.cv_results_['std_test_score']
    params = grid_result.cv_results_['params']
    for mean, stdev, param in zip(means, stds, params):
      print("%f (%f) with: %r" % (mean, stdev, param))
    print()

    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    y_true, y_pred = y_te, grid_search.predict(X_te)
    print(classification_report(y_true, y_pred))
    print()

  

# Tuning hyper-parameters for precision

Best: 0.905526 using {'learning_rate': 1, 'n_estimators': 500, 'subsample': 0.9}
Grid scores on development set:

0.500000 (0.000000) with: {'learning_rate': 0.001, 'n_estimators': 100, 'subsample': 0.5}
0.500000 (0.000000) with: {'learning_rate': 0.001, 'n_estimators': 100, 'subsample': 0.9}
0.516936 (0.003893) with: {'learning_rate': 0.001, 'n_estimators': 500, 'subsample': 0.5}
0.515067 (0.006117) with: {'learning_rate': 0.001, 'n_estimators': 500, 'subsample': 0.9}
0.721764 (0.013967) with: {'learning_rate': 0.001, 'n_estimators': 750, 'subsample': 0.5}
0.722389 (0.013178) with: {'learning_rate': 0.001, 'n_estimators': 750, 'subsample': 0.9}
0.734126 (0.012945) with: {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.5}
0.735381 (0.013033) with: {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.9}
0.897361 (0.048077) with: {'learning_rate': 1, 'n_estimators': 100, 'subsample': 0.5}
0.884088 (0.071025) with: {'learning_

In [36]:
grid = dict()
grid['n_estimators'] = [ 500]
grid['learning_rate'] = [0.1]
grid['subsample'] = [0.9]
scores = ["precision","recall"]
# define the model with default hyperparameters
model = GradientBoostingClassifier()
# define the evaluation procedure
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

# define the grid search procedure
for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring="%s_macro" % score)
    # execute the grid search
    grid_result = grid_search.fit(X_tr, y_tr)
    # summarize the best score and configuration
    print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
    # summarize all scores that were evaluated
    print("Grid scores on development set:")
    print()
    means = grid_result.cv_results_['mean_test_score']
    stds = grid_result.cv_results_['std_test_score']
    params = grid_result.cv_results_['params']
    for mean, stdev, param in zip(means, stds, params):
      print("%f (%f) with: %r" % (mean, stdev, param))
    print()

    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    y_true, y_pred = y_te, grid_search.predict(X_te)
    print(classification_report(y_true, y_pred))
    print()

  

# Tuning hyper-parameters for precision

Best: 0.972355 using {'learning_rate': 0.1, 'n_estimators': 500, 'subsample': 0.9}
Grid scores on development set:

0.972355 (0.005689) with: {'learning_rate': 0.1, 'n_estimators': 500, 'subsample': 0.9}

Detailed classification report:

The model is trained on the full development set.
The scores are computed on the full evaluation set.

              precision    recall  f1-score   support

           0       0.97      0.88      0.92      2153
           1       0.98      1.00      0.99     12123

    accuracy                           0.98     14276
   macro avg       0.98      0.94      0.95     14276
weighted avg       0.98      0.98      0.98     14276


# Tuning hyper-parameters for recall

Best: 0.932447 using {'learning_rate': 0.1, 'n_estimators': 500, 'subsample': 0.9}
Grid scores on development set:

0.932447 (0.008167) with: {'learning_rate': 0.1, 'n_estimators': 500, 'subsample': 0.9}

Detailed classification report:

The model is t